bson/json_util.py

```diff
@@ -311,6 +311,28 @@ def _arguments_repr(self):
                     self.json_mode,
                     super(JSONOptions, self)._arguments_repr()))
 
+    def with_options(self, **kwargs):
+        """
+        Make a copy of this JSONOptions, overriding some options::
+
+            >>> from bson.json_util import CANONICAL_JSON_OPTIONS
+            >>> CANONICAL_JSON_OPTIONS.tz_aware
+            True
+            >>> json_options = CANONICAL_JSON_OPTIONS.with_options(tz_aware=False)
+            >>> json_options.tz_aware
+            False
+
+        .. versionadded:: 3.12
+        """
+        return JSONOptions(
+            strict_number_long=kwargs.pop(
+                'strict_number_long', self.strict_number_long),
+            datetime_representation=kwargs.pop(
+                'datetime_representation', self.datetime_representation),
+            strict_uuid=kwargs.pop('strict_uuid', self.strict_uuid),
+            json_mode=kwargs.pop('json_mode', self.json_mode),
+            **kwargs)
```

Comment:
I think we need to copy all the JSONOptions and CodecOptions attributes here otherwise the some settings will be lost.

Reply:
Good point. I made this change and also added a test case that fails if this is not true.